pkg/util/validation/validate.go

```diff
@@ -68,51 +69,65 @@ const (
 	ExemplarMaxLabelSetLength = 128
 )
 
-// DiscardedSamples is a metric of the number of discarded samples, by reason.
-var DiscardedSamples = prometheus.NewCounterVec(
-	prometheus.CounterOpts{
-		Name: "cortex_discarded_samples_total",
-		Help: "The total number of samples that were discarded.",
-	},
-	[]string{discardReasonLabel, "user"},
-)
+type ValidateMetrics struct {
+	DiscardedSamples   *prometheus.CounterVec
+	DiscardedExemplars *prometheus.CounterVec
+	DiscardedMetadata  *prometheus.CounterVec
+}
 
-// DiscardedExemplars is a metric of the number of discarded exemplars, by reason.
-var DiscardedExemplars = prometheus.NewCounterVec(
-	prometheus.CounterOpts{
-		Name: "cortex_discarded_exemplars_total",
-		Help: "The total number of exemplars that were discarded.",
-	},
-	[]string{discardReasonLabel, "user"},
-)
+func registerCollector(r prometheus.Registerer, c prometheus.Collector) {
+	err := r.Register(c)
+	if err != nil && !errors.As(err, &prometheus.AlreadyRegisteredError{}) {
+		panic(err)
+	}
+}
 
-// DiscardedMetadata is a metric of the number of discarded metadata, by reason.
-var DiscardedMetadata = prometheus.NewCounterVec(
-	prometheus.CounterOpts{
-		Name: "cortex_discarded_metadata_total",
-		Help: "The total number of metadata that were discarded.",
-	},
-	[]string{discardReasonLabel, "user"},
-)
+func NewValidateMetrics(r prometheus.Registerer) *ValidateMetrics {
+	discardedSamples := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "cortex_discarded_samples_total",
+			Help: "The total number of samples that were discarded.",
+		},
+		[]string{discardReasonLabel, "user"},
+	)
+	registerCollector(r, discardedSamples)
+	discardedExemplars := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "cortex_discarded_exemplars_total",
+			Help: "The total number of exemplars that were discarded.",
+		},
+		[]string{discardReasonLabel, "user"},
+	)
+	registerCollector(r, discardedExemplars)
+	discardedMetadata := prometheus.NewCounterVec(
```

Comment:
I wonder if we can just use `promauto` package here which automatically registers the collector.

Reply:
We cannot because `promauto` does `MustRegister` which will panic if there is a registration error (this includes `AlreadyRegisteredError`).